Request: exhale - (USAC, xHE-AAC) encoder

Wonder if you could add this new (USAC, xHE-AAC) encoder to mabs

https://gitlab.com/ecodis/exhale

https://gitlab.com/ecodis/exhale/-/blob/master/README.md

Doesn't compile directly with gcc on MinGW.

```
H:/repos/media-autobuild_suite/build/exhale-git/src/app/../../src/app/exhaleApp.cpp: In function 'int main(int, char**)':
H:/repos/media-autobuild_suite/build/exhale-git/src/app/../../src/app/exhaleApp.cpp:246:85: error: '_SH_DENYWR' was not declared in this scope
  246 |     if (_sopen_s (&inFileHandle, inFileName, _O_RDONLY | _O_SEQUENTIAL | _O_BINARY, _SH_DENYWR, _S_IREAD) != 0)
      |                                                                                     ^~~~~~~~~~
H:/repos/media-autobuild_suite/build/exhale-git/src/app/../../src/app/exhaleApp.cpp:320:100: error: '_SH_DENYRD' was not declared in this scope
  320 |     if (_sopen_s (&outFileHandle, outFileName, i | _O_SEQUENTIAL | _O_CREAT | _O_EXCL | _O_BINARY, _SH_DENYRD, _S_IWRITE) != 0)
      |                                                                                                    ^~~~~~~~~~
make[1]: *** [../makefile.base:112: ../../build/exhaleApp.d.o] Error 1
make[1]: Leaving directory '/build/exhale-git/src/app'
make: *** [makefile:18: all] Error 2
```
Looking into this.

--- a/build/media-suite_compile.sh
+++ b/build/media-suite_compile.sh
@@ -710,11 +710,11 @@ fi
 
 _check=(bin-audio/exhale.exe)
 if [[ $standalone = y && $exhale = y ]] &&
-    do_vcs "https://gitlab.com/ecodis/exhale"; then
-	do_patch "https://gist.githubusercontent.com/moisespr123/abac078f080cca10d605ea0390323bc1/raw/exhale-patch"
+    do_vcs "https://gitlab.com/ecodis/exhale.git"; then
+    do_patch "https://gist.githubusercontent.com/moisespr123/abac078f080cca10d605ea0390323bc1/raw/exhale-patch"
     do_uninstall "${_check[@]}"
-	do_make release
-	do_install bin/exhale.exe bin-audio/
+    do_make release
+    do_install bin/exhale.exe bin-audio/
     do_checkIfExist
 fi
 

--- a/build/media-suite_compile.sh
+++ b/build/media-suite_compile.sh
@@ -709,7 +709,7 @@ if [[ $standalone = y && $faac = y ]] && ! files_exist "${_check[@]}" &&
 fi
 
 _check=(bin-audio/exhale.exe)
-if [[ $standalone = y && $exhale = y ]] &&
+if [[ $exhale = y ]] &&
     do_vcs "https://gitlab.com/ecodis/exhale.git"; then
     do_patch "https://gist.githubusercontent.com/moisespr123/abac078f080cca10d605ea0390323bc1/raw/exhale-patch"
     do_uninstall "${_check[@]}"

--- a/build/media-suite_compile.sh
+++ b/build/media-suite_compile.sh
@@ -711,9 +711,8 @@ fi
 _check=(bin-audio/exhale.exe)
 if [[ $exhale = y ]] &&
     do_vcs "https://gitlab.com/ecodis/exhale.git"; then
-    do_patch "https://gist.githubusercontent.com/moisespr123/abac078f080cca10d605ea0390323bc1/raw/exhale-patch"
     do_uninstall "${_check[@]}"
-    do_make release
+    do_make release ADDITIONAL_LDFLAGS="$LDFLAGS"
     do_install bin/exhale.exe bin-audio/
     do_checkIfExist
 fi

--- a/build/media-suite_compile.sh
+++ b/build/media-suite_compile.sh
@@ -47,6 +47,7 @@ while true; do
     --bmx=* ) bmx=${1#*=} && shift ;;
     --aom=* ) aom=${1#*=} && shift ;;
     --faac=* ) faac=${1#*=} && shift ;;
+    --exhale=* ) exhale=${1#*=} && shift ;;
     --ffmbc=* ) ffmbc=${1#*=} && shift ;;
     --curl=* ) curl=${1#*=} && shift ;;
     --cyanrip=* ) cyanrip=${1#*=} && shift ;;
@@ -704,6 +705,16 @@ if [[ $standalone = y && $faac = y ]] && ! files_exist "${_check[@]}" &&
     extracommands=()
     [[ $standalone = n ]] && extracommands+=(--disable-frontend)
     do_separate_confmakeinstall audio "${extracommands[@]}"
+    do_checkIfExist
+fi
+
+_check=(bin-audio/exhale.exe)
+if [[ $standalone = y && $exhale = y ]] &&
+    do_vcs "https://gitlab.com/ecodis/exhale"; then
+	do_patch "https://gist.githubusercontent.com/moisespr123/abac078f080cca10d605ea0390323bc1/raw/exhale-patch"
+    do_uninstall "${_check[@]}"
+	do_make release
+	do_install bin/exhale.exe bin-audio/
     do_checkIfExist
 fi
 
@@ -2392,7 +2403,7 @@ if [[ $cyanrip = y ]]; then
 fi
 
 if [[ $vlc == y ]]; then
-    do_pacman_install lib{nfs,shout,samplerate,microdns,secret} \
+    do_pacman_install lib{cddb,nfs,shout,samplerate,microdns,secret} \
         a52dec taglib gtk3 lua perl
 
     # Remove useless shell scripts file that causes errors when stdout is not a tty.
@@ -2590,7 +2601,8 @@ if [[ $vlc == y ]]; then
         # msys2's patches
         # Issues due to conflicting `vlc_module_name` between libvlc and libvlccore when linking vlc-static.exe and undefines.
         # having gpg-error after GCRYPT_LIBS causes some issues, and since it's already included in GCRYPT_LIBS
-        do_patch "https://gist.githubusercontent.com/1480c1/8c50a0867aa1afceac064d2162120dde/raw/vlc-mabs.patch" am
+        do_patch "https://gist.githubusercontent.com/moisespr123/2369978dc603ed1e67bdf7aba7304416/raw/vlc-corrected-patch" am
+        do_patch "https://gist.githubusercontent.com/moisespr123/70e16f70f5d016c3d0f0dacbf97526f0/raw/medialib-patch"
 
         do_autoreconf
         # All of the disabled are because of multiple issues both on the installed libs and on vlc's side.

--- a/media-autobuild_suite.bat
+++ b/media-autobuild_suite.bat
@@ -135,9 +135,9 @@ set mpv_options_full=dvdnav cdda #egl-angle #html-build ^
 
 set iniOptions=msys2Arch arch license2 vpx2 x2643 x2652 other265 flac fdkaac mediainfo ^
 soxB ffmpegB2 ffmpegUpdate ffmpegChoice mp4box rtmpdump mplayer2 mpv cores deleteSource ^
-strip pack logging bmx standalone updateSuite aom faac ffmbc curl cyanrip2 redshift rav1e ^
-ripgrep dav1d vvc jq dssim avs2 timeStamp noMintty ccache svthevc svtav1 svtvp9 xvc jo ^
-vlc CC
+strip pack logging bmx standalone updateSuite aom faac exhale ffmbc curl cyanrip2 redshift ^
+rav1e ripgrep dav1d vvc jq dssim avs2 timeStamp noMintty ccache svthevc svtav1 svtvp9 xvc ^
+jo vlc CC
 
 set deleteIni=0
 set ini=%build%\media-autobuild_suite.ini
@@ -590,6 +590,26 @@ if %buildfaac%==1 set "faac=y"
 if %buildfaac%==2 set "faac=n"
 if %buildfaac% GTR 2 GOTO faac
 if %deleteINI%==1 echo.faac=^%buildfaac%>>%ini%
+
+:exhale
+if %exhaleINI%==0 (
+    echo -------------------------------------------------------------------------------
+    echo -------------------------------------------------------------------------------
+    echo.
+    echo. Build exhale binary? [open-source ISO/IEC 23003-3 USAC, xHE-AAC encoder]
+    echo. 1 = Yes
+    echo. 2 = No
+    echo.
+    echo -------------------------------------------------------------------------------
+    echo -------------------------------------------------------------------------------
+    set /P buildexhale="Build exhale: "
+) else set buildexhale=%exhaleINI%
+
+if "%buildexhale%"=="" GOTO exhale
+if %buildexhale%==1 set "exhale=y"
+if %buildexhale%==2 set "exhale=n"
+if %buildexhale% GTR 2 GOTO exhale
+if %deleteINI%==1 echo.exhale=^%buildexhale%>>%ini%
 
 :mediainfo
 if %mediainfoINI%==0 (
@@ -1671,8 +1691,8 @@ set compileArgs=--cpuCount=%cpuCount% --build32=%build32% --build64=%build64% ^
 --other265=%other265% --flac=%flac% --fdkaac=%fdkaac% --mediainfo=%mediainfo% --sox=%sox% ^
 --ffmpeg=%ffmpeg% --ffmpegUpdate=%ffmpegUpdate% --ffmpegChoice=%ffmpegChoice% --mplayer=%mplayer% ^
 --mpv=%mpv% --license=%license2%  --stripping=%stripFile% --packing=%packFile% --rtmpdump=%rtmpdump% ^
---logging=%logging% --bmx=%bmx% --standalone=%standalone% --aom=%aom% --faac=%faac% --ffmbc=%ffmbc% ^
---curl=%curl% --cyanrip=%cyanrip% --redshift=%redshift% --rav1e=%rav1e% --ripgrep=%ripgrep% ^
+--logging=%logging% --bmx=%bmx% --standalone=%standalone% --aom=%aom% --faac=%faac% --exhale=%exhale% ^
+--ffmbc=%ffmbc% --curl=%curl% --cyanrip=%cyanrip% --redshift=%redshift% --rav1e=%rav1e% --ripgrep=%ripgrep% ^
 --dav1d=%dav1d% --vvc=%vvc% --jq=%jq% --jo=%jo% --dssim=%dssim% --avs2=%avs2% --timeStamp=%timeStamp% ^
 --noMintty=%noMintty% --ccache=%ccache% --svthevc=%svthevc% --svtav1=%svtav1% --svtvp9=%svtvp9% --xvc=%xvc% ^
 --vlc=%vlc%

--- a/media-autobuild_suite.bat
+++ b/media-autobuild_suite.bat
@@ -600,6 +600,8 @@ if %exhaleINI%==0 (
     echo. 1 = Yes
     echo. 2 = No
     echo.
+    echo. Binaries being built do not depend on "standalone=y"
+    echo.
     echo -------------------------------------------------------------------------------
     echo -------------------------------------------------------------------------------
     set /P buildexhale="Build exhale: "

